evennia/contrib/auditing.py

```diff
@@ -0,0 +1,167 @@
+private communications.
+***************
+
+Installation:
+
+Designate this class as the SERVER_SESSION_CLASS in `settings.py`.
+
+i.e. SERVER_SESSION_CLASS = "evennia.contrib.auditing.AuditedServerSession"
+
+There are three additional settings that govern operation-- whether you
+want to record inbound strings and/or outbound strings, and what method
+you want the data sent to.
+
+AUDIT_METHOD = just_write_it
+AUDIT_IN = True or False
+AUDIT_OUT = True or False
+
+Recording outbound strings can be noisy-- a single message broadcast to 12
+occupants of a room will generate 12 of the same message, one to each.
+"""
+
+import os
+import socket
+
+from django.utils import timezone
+from evennia import settings as ev_settings
+from evennia.utils import logger
+from evennia.server.serversession import ServerSession
+
+class AuditedServerSession(ServerSession):
+    def __init__(self, *args, **kwargs):
+        super(AuditedServerSession, self).__init__(*args, **kwargs)
+        
+        # Attributes governing auditing of commands and where to send logs
+        self.audit_method = getattr(ev_settings, 'AUDIT_METHOD', None)
+        self.audit_in = getattr(ev_settings, 'AUDIT_IN', False)
+        self.audit_out = getattr(ev_settings, 'AUDIT_OUT', False)
+    
+    def audit(self, **kwargs):
+        """
+        Creates a log entry from the given session.
+        """
+        # Sanitize user input
+        session = self
+        src = kwargs.get('src', '?')
+        
+        if src == 'client':
+            try:
+                data = kwargs['text'][0][0].strip()
+            except IndexError:
+                logger.log_err("Failed to log %s" % kwargs['text'])
+                return False
+                
+        elif src == 'server':
+            try:
+                data = kwargs['text'].strip()
+            except:
+                data = str(kwargs)
+            
+        # Do not log empty lines
+        if not data: return False
+        
+        # Get current session's IP address
+        ip_token = session.address
+        
+        # Capture Account name and dbref together
+        account = None
+        account = session.get_account()
+        account_token = 'null'
+        if account: 
+            account_token = '%s%s' % (account.key, account.dbref)
+            
+        # Capture Character name and dbref together
+        char = None
+        char = session.get_puppet()
+        char_token = 'null'
+        if char:
+            char_token = '%s%s' % (char.key, char.dbref)
+            
+        # Capture Room name and dbref together
+        room = None
+        room_token = 'null'
+        if char:
+            room = char.location
+            room_token = '%s%s' % (room.key, room.dbref)
+            
+        # Compile the IP, Account, Character, Room, and the message.
+        parsed = {
+            'time': str(timezone.now()),
+            'hostname': socket.getfqdn(),
```

Comment:
This info, plus application and pid appears identical for all sends. Is it necessary to store with every log?

Reply:
Nope.

Time should not be identical in more than a few sends at a time-- it should resolve to millisecond accuracy. I'm not sure if that was meant to be in the scope of your question.

Hostname is identical most of the time in most deployments but it *will* change under certain containerized architectures. It can be useful to know on what node the event was recorded (especially in cases where logs should be indexed/bucketed by node name instead of date, or stored locally before forwarding).

A lot of the other metadata (appame, pid) are provided for the sake of facilitating output in syslog format, interest in which was brought up by whitenoise in previous discussion. Same with object references as per tekmunkey.

Compressed, text storage is cheap and having too much data is better than not having enough when the need arises to conduct a review. There's no point in logging anything if the logs are too incomplete to compose a narrative. These considerations are why I ended up just using dicts as a medium instead of passing strings-- if deduplication or shaving every last byte is a concern, the op can easily cherry-pick what they want from the data buffet.

Reply:
Time was clearly different, but if you say the host name can change in containerization, that's fair enough. I know syslog format was discussed before; I was not aware PID was necessary for this. That's ok then. :)